### PR TITLE
Add pre-commit and its dependencies

### PR DESCRIPTION
**Pull request by jfroche**
pre-commit is "a framework for managing and maintaining multi-language
pre-commit hooks". More info here: https://pre-commit.com

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Please split up these changes into multiple commits.

Precisely like:

- add yourself to the maintainers list
```
maintainers: add name
```

- one each for package initialization 
```
pkg-name: init at version
```
Also preferably with its dependents preceding pre-commit.

##### More can be seen in [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md)

File: pkgs/development/python-modules/importlib-metadata/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+, fetchPypi
+, lib
+, recursivePthLoader
+, isPy3k
+}:
+
+buildPythonPackage rec {
+  pname = "importlib_metadata";
+  version = "0.6";
+  name = "importlib_metadata";
```
Comment (dotlambda) [inline pkgs/development/python-modules/importlib-metadata/default.nix]: drop this line

File: pkgs/development/python-modules/importlib-metadata/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  disabled = !isPy3k;
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "074v6mq5dbbr0y697fgajm503sg83faxzvwz41jxy6h0z622rc1n";
+  };
+
+  pythonPath = [ recursivePthLoader ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/importlib-metadata/default.nix]: Why is this needed? Why don't you use `propagatedBuildInputs` or `nativeBuildInputs`?

File: pkgs/development/python-modules/importlib-metadata/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/development/python-modules/importlib-metadata/default.nix]: Why? Add a comment.

File: pkgs/development/python-modules/importlib-metadata/default.nix [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  propagatedBuildInputs = [ ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/importlib-metadata/default.nix]: remove

File: pkgs/development/python-modules/importlib-resources/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+, fetchPypi
+, lib
+, recursivePthLoader
+, isPy3k
+}:
+
+buildPythonPackage rec {
+  pname = "importlib_resources";
+  version = "1.0.1";
+  name = "importlib_resources";
```
Comment (dotlambda) [inline pkgs/development/python-modules/importlib-resources/default.nix]: remove

File: pkgs/development/python-modules/importlib-resources/default.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1yrf2nyrxigi3f6h4kj7sqxzb4dwkqgqrcb2saprn55ccbh59x3k";
+  };
+
+  disabled = !isPy3k;
+
+  pythonPath = [ recursivePthLoader ];
+  doCheck = false;
+  propagatedBuildInputs = [ ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/importlib-resources/default.nix]: same as above

File: pkgs/development/python-modules/pre-commit/default.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+, importlib-metadata
+, toml
+, cfgv
+, python36
+}:
+
+buildPythonPackage rec {
+  pname = "pre_commit";
+  version = "1.12.0";
+  name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pre-commit/default.nix]: drop this line

File: pkgs/development/python-modules/pre-commit/default.nix [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+, six
+, cached-property
+, importlib-resources
+, importlib-metadata
+, toml
+, cfgv
+, python36
+}:
+
+buildPythonPackage rec {
```
Comment (dotlambda) [inline pkgs/development/python-modules/pre-commit/default.nix]: Is this a library or an application? Please read https://nixos.org/nixpkgs/manual/#building-packages-and-applications.

File: pkgs/development/python-modules/pre-commit/default.nix [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  version = "1.12.0";
+  name = "${pname}-${version}";
+  format = "wheel";
+
+  src = fetchPypi {
+    inherit pname version format;
+    sha256 = "0q95s22vs0jcbr7k9i0qgq7s30pfcicjkad0bqblb1y5w65bshkm";
+  };
+
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/development/python-modules/pre-commit/default.nix]: Why?

File: pkgs/development/python-modules/pre-commit/default.nix [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+, nodeenv
+, virtualenv
+, aspy-yaml
+, six
+, cached-property
+, importlib-resources
+, importlib-metadata
+, toml
+, cfgv
+, python36
```
Comment (dotlambda) [inline pkgs/development/python-modules/pre-commit/default.nix]: Do not import a specific version!

File: pkgs/development/python-modules/pre-commit/default.nix [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  postInstall = ''
+    export PYTHONPATH="$out/lib/${python36.libPrefix}/site-packages:$PYTHONPATH"
+    for i in "$out/bin/"*
+    do
+      wrapProgram "$i" --set PYTHONPATH $PYTHONPATH
+    done
+  '';
+
+  meta = {
+    description = "A framework for managing and maintaining multi-language pre-commit hooks.";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pre-commit/default.nix]: No period: https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

File: pkgs/top-level/python-packages.nix [hunk 12]
```diff
@@ -11054,6 +11054,11 @@ EOF
     inherit python;
   })).python;
 
+  importlib-metadata = callPackage ../development/python-modules/importlib-metadata { };
+  importlib-resources = callPackage ../development/python-modules/importlib-resources { };
+  setuptools-markdown = callPackage ../development/python-modules/setuptools-markdown { };
+  pre-commit = callPackage ../development/python-modules/pre-commit { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: This file should be sorted alphabetically.

File: pkgs/development/python-modules/setuptools-markdown/default.nix [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  name = "setuptools_markdown";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0xzx2krdn2g1kw949mvxkxkc1gflpqmrj9lxdnvdpds3wds66nh6";
+  };
+
+  propagatedBuildInputs = [ pypandoc ];
+  pythonPath = [ recursivePthLoader ];
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/development/python-modules/setuptools-markdown/default.nix]: same as above

File: pkgs/development/python-modules/setuptools-markdown/default.nix [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+, fetchPypi
+, lib
+, recursivePthLoader
+, pypandoc
+}:
+
+buildPythonPackage rec {
+  pname = "setuptools-markdown";
+  version = "0.2";
+  name = "setuptools_markdown";
```
Comment (dotlambda) [inline pkgs/development/python-modules/setuptools-markdown/default.nix]: drop this line

File: pkgs/development/python-modules/setuptools-markdown/default.nix [hunk 15]
```diff
@@ -0,0 +1,27 @@
+{ buildPythonPackage
+, fetchPypi
+, lib
+, recursivePthLoader
+, pypandoc
+}:
+
+buildPythonPackage rec {
+  pname = "setuptools-markdown";
```
Comment (dotlambda) [inline pkgs/development/python-modules/setuptools-markdown/default.nix]: Are you sure that this works? Pypandoc is broken.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by dotlambda** [comment]
I only found out after reviewing that this package already exists: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/git-and-tools/pre-commit/default.nix

**Pull request closed without merging**